### PR TITLE
JDK-8293313: NMT: Rework MallocLimit

### PR DESCRIPTION
**Pull request by tstuefe**
Rework NMT `MallocLimit` to make it more useful for native OOM analysis. Also remove `MallocMaxTestWords`, which had been mostly redundant with NMT (for background, see JDK-8293292).

### Background

Some months ago we had [JDK-8291919](https://bugs.openjdk.org/browse/JDK-8291919), a compiler regression that caused compiler arenas to grow boundlessly. Process was usually OOM-killed before we could take a look, so this was difficult to analyze.

To improve analysis options, we added NMT *malloc limits* with [JDK-8291878](https://bugs.openjdk.org/browse/JDK-8291878). Malloc limits let us set one or multiple limits to malloc load. Surpassing a limit causes the VM to stop with a fatal error in the allocation function, giving us a hs-err file and a core right at the point that (probably) causes the leak. This makes error analysis a lot simpler, and is also valuable for regression-testing footprint usage.

Some people wished for a way to not end with a fatal error but to mimic a native OOM (causing os::malloc to return NULL). This would allow us to test resilience in the face of native OOMs, among other things.

### Patch

- Expands the `MallocLimit` option syntax, allowing for an "oom mode" that mimics an oom:

```
Global form:
-XX:MallocLimit=<size>[:<mode>]
Category-specific form:
-XX:MallocLimit=<category>:<size>[:<mode>][,<category>:<size>[:<mode>] ...]
Examples:
-XX:MallocLimit=3g
-XX:MallocLimit=3g:oom
-XX:MallocLimit=compiler:200m:oom
```

- moves parsing of `-XX:MallocLimit` out of arguments.cpp into `mallocLimit.hpp/cpp`, and rewrites it.

- changes when limits are checked. Before, the limits were checked *after* the allocation that caused the threshold overflow. Now we check beforehand.

- removes `MallocMaxTestWords`, replacing its uses with `MallocLimit`

- adds new gtests and new jtreg tests

- removes a bunch of jtreg tests which are now better served by the new gtests.

- gives us more precise error messages upon reaching a limit. For example:

before:
```
#  fatal error: MallocLimit: category "Compiler" reached limit (size: 20997680, limit: 20971520)
```

now:
```
#  fatal error: MallocLimit: reached category "Compiler" limit (triggering allocation size: 1920B, allocated so far: 20505K, limit: 20480K)
```

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issues
 * [JDK-8293313](https://bugs.openjdk.org/browse/JDK-8293313): NMT: Rework MallocLimit
 * [JDK-8293292](https://bugs.openjdk.org/browse/JDK-8293292): Remove MallocMaxTestWords

### Reviewers
 * [Johan Sjölen](https://openjdk.org/census#jsjolen) (@jdksjolen - Committer)
 * [Gerard Ziemski](https://openjdk.org/census#gziemski) (@gerard-ziemski - Committer) ⚠️ Review applies to [7e9c2617](https://git.openjdk.org/jdk/pull/11371/files/7e9c26179f04886ebaec46a353f894f62cf2f9b4)
 * [Lutz Schmidt](https://openjdk.org/census#lucy) (@RealLucy - **Reviewer**)
 * [Matthias Baesken](https://openjdk.org/census#mbaesken) (@MBaesken - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11371/head:pull/11371` \
`$ git checkout pull/11371`

Update a local copy of the PR: \
`$ git checkout pull/11371` \
`$ git pull https://git.openjdk.org/jdk pull/11371/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11371`

View PR using the GUI difftool: \
`$ git pr show -t 11371`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11371.diff">https://git.openjdk.org/jdk/pull/11371.diff</a>

</details>

### CODE REVIEW

**Comment by gerard-ziemski** [comment]
Just FYI I'm taking a look at this...

File: test/hotspot/gtest/nmt/test_nmt_malloclimit.cpp
Comment (jdksjolen) [inline test/hotspot/gtest/nmt/test_nmt_malloclimit.cpp:57]: Seems like a left over from development.

File: test/hotspot/gtest/nmt/test_nmt_malloclimit.cpp
Comment (jdksjolen) [inline test/hotspot/gtest/nmt/test_nmt_malloclimit.cpp:60]: Using EXPECT would let the test proceed, is it necessary to use ASSERT here?

File: test/hotspot/gtest/nmt/test_nmt_malloclimit.cpp
Comment (jdksjolen) [inline test/hotspot/gtest/nmt/test_nmt_malloclimit.cpp:82]: >malloclimit_per_category

Can we use CamelCase for test case names instead? This is a convention that we *should* use for gtest: http://google.github.io/googletest/faq.html

Unfortunately, we don't consistently do this, but I do want to avoid introducing new instances of this.

File: test/hotspot/jtreg/runtime/Unsafe/Reallocate.java
Comment (jdksjolen) [inline test/hotspot/jtreg/runtime/Unsafe/Reallocate.java:65]: Why are we changing the test here? Just because 800m is a lot larger than the required 100m?

**Review by jdksjolen — COMMENTED** [review]
This looks good. I've added some nits for style and a couple of comments concerning style which aren't nits. Logic wise, easy to follow and straight forward code, I could find nothing to complain about.

File: src/hotspot/share/services/mallocLimit.hpp
Comment (tstuefe) [reply src/hotspot/share/services/mallocLimit.hpp:33]: Well, `MallocLimit` would class with the option name. I'll think of something.

File: src/hotspot/share/services/mallocLimit.hpp
Comment (tstuefe) [reply src/hotspot/share/services/mallocLimit.hpp:40]: Sure

File: src/hotspot/share/services/mallocLimit.cpp
Comment (tstuefe) [reply src/hotspot/share/services/mallocLimit.cpp:43]: It means end of input, eol. The ParserHelper is dumb, it does not know what successful parsing is. It just extracts stuff.

File: test/hotspot/gtest/nmt/test_nmt_malloclimit.cpp
Comment (tstuefe) [reply test/hotspot/gtest/nmt/test_nmt_malloclimit.cpp:57]: I often leave these in at neuralgic points to quickly analyze problems, they use the LOG_HERE macro in testutils.hpp. But I can remove it.

File: test/hotspot/gtest/nmt/test_nmt_malloclimit.cpp
Comment (tstuefe) [reply test/hotspot/gtest/nmt/test_nmt_malloclimit.cpp:82]: I can see the point, though my names often follow a vague hierarchy if the suitename is already very generic. E.g. "malloclimit_xxx". - suitename is "NMT" and should remain so, all tests in this file concern MallocLimit, and then something. I like to separate "MallocLimit" visually from the "something".

I guess its okay, just a minor annoyance.

File: test/hotspot/jtreg/runtime/Unsafe/Reallocate.java
Comment (tstuefe) [reply test/hotspot/jtreg/runtime/Unsafe/Reallocate.java:65]: MallocMaxTestWords took the value in *words*, that's where the 8 comes in. Note that it was broken anyway since it never shrank - it ignored free calls. Therefore the values given to MallocMaxTestWords have only ever been guesswork.

File: src/hotspot/share/services/mallocLimit.cpp
Comment (tstuefe) [reply src/hotspot/share/services/mallocLimit.cpp]: > This looks good. I've added some nits for style and a couple of comments concerning style which aren't nits. Logic wise, easy to follow and straight forward code, I could find nothing to complain about.

Thank you Johan, 

I massaged in your code feedback.

Cheers, Thomas

**Comment by tstuefe** [comment]
> This looks good. I've added some nits for style and a couple of comments concerning style which aren't nits. Logic wise, easy to follow and straight forward code, I could find nothing to complain about.

Thanks @jdksjolen,

I massaged in your feedback.

Cheers, Thomas

File: src/hotspot/share/runtime/os.cpp
Comment (gerard-ziemski) [inline src/hotspot/share/runtime/os.cpp:707]: Can we make it a bit easier to parse by rewriting it as:

`    if ((size > old_size) && MemTracker::check_exceeds_limit(size - old_size, memflags)) {`

To me personally that is easier to read.

I see a bigger issue here, however. In the worse case scenario where the os is unable to expand the memory chunk in place, it will need to allocate `size` bytes in a new region, so the total **potential** resources from the os point of view is `size+old_size` bytes. Shouldn't we assume this worse case and test for that?

I'd rather get some false positives that miss a single true positive... 

And if that were the case we wouldn't need to check for `(size > old_size) ` and we should have just:

`    if (MemTracker::check_exceeds_limit(size + old_size, memflags)) {`

with a comment explaining this choice.

**Review by gerard-ziemski — CHANGES_REQUESTED** [review]
Still reviewing, just wanted to share this potential issue right away...

File: src/hotspot/share/services/mallocTracker.inline.hpp
Comment (gerard-ziemski) [inline src/hotspot/share/services/mallocTracker.inline.hpp:56]: I would prefer to see `suppress_limit_handling()` checked inside `total_limit_reached()`, same for `category_limit_reached()`. This way we would force those APIs to always obey the suppression without having to bother to add `suppress_limit_handling()`, they could call `VMError::is_error_reported()` directly.

**Review by gerard-ziemski — CHANGES_REQUESTED** [review]
Made more progress on the review today, tomorrow I have the parser to review and java tests...

File: src/hotspot/share/runtime/os.cpp
Comment (tstuefe) [reply src/hotspot/share/runtime/os.cpp:707]: > Can we make it a bit easier to parse by rewriting it as:
> 
> ` if ((size > old_size) && MemTracker::check_exceeds_limit(size - old_size, memflags)) {`
> 
> To me personally that is easier to read.
> 

Sure

> I see a bigger issue here, however. In the worst case scenario, where the os is unable to expand the memory chunk in place, it will need to allocate `size` bytes in a new region, so the total **potential** resources from the os point of view is `size+old_size` bytes for that particular allocation. Shouldn't we assume this worst case and test for that? I.e.
> 
> ` if ((size > old_size) && MemTracker::check_exceeds_limit(size, memflags)) {`
> 
> We would need a comment here explaining this choice if we make this change.
> 
> I'd rather get some false positives that miss a single true positive...

Nah, no need to overthink this. 

If you try to predict how much memory your malloc causes your process to allocate, it gets very fuzzy very quickly:

Below are at least two allocating layers (libc and the kernel's vm manager). Both of them balance memory and CPU footprint and allocation speed. No libc or kernel is the same, too. Each layer below you may return cached or partly cached memory and apply an often sizable overhead to what you are allocating. So our mallocs and frees have only a very indirect effect on RSS. And we don't even see all mallocs here, just the ones from libjvm.

Therefore it is best to limit the meaning of MallocLimit to "limit to how much hotspot is allowed to allocate". That is also the most useful. For limiting the memory of a process, different os-side tools exist (cgroup limits, ulimit, etc).

File: src/hotspot/share/runtime/globals.hpp
Comment (tstuefe) [reply src/hotspot/share/runtime/globals.hpp:1344]: I prefer MallocLimit, short and snappy. And the fact that NMT is supervising the limit is an implementation detail. It used to be different with MallocMaxTestWords and may be different in the future.

File: src/hotspot/share/services/mallocTracker.inline.hpp
Comment (tstuefe) [reply src/hotspot/share/services/mallocTracker.inline.hpp:35]: > Why did we bother to wrap `VMError::is_error_reported()` into `suppress_limit_handling()`?
> 

Because during error handling, code may malloc() too (bad practice, but it can happen). If it does, I don't want circular assertions to fire; I want a clean, complete hs-err file.

> Are you anticipating more exclusions here in the future?

None I can think of.

File: src/hotspot/share/services/mallocTracker.inline.hpp
Comment (tstuefe) [reply src/hotspot/share/services/mallocTracker.inline.hpp:56]: > I would prefer to see `suppress_limit_handling()` checked inside `total_limit_reached()`, same for `category_limit_reached()`. This way we would force those APIs to always obey the suppression without having to bother to add `suppress_limit_handling()`, they could call `VMError::is_error_reported()` directly.

Note that `total_limit_reached()` and `category_limit_reached()` are private APIs. 

If we move `suppress_limit_handling()` into `total_limit_reached()`, `total_limit_reached()` would need to report it back to `MallocMemorySummary::check_exceeds_limit` since its processed there too, it affects the return code. So `total_limit_reached()` cannot be void anymore, and gets somewhat more complex.

**Comment by tstuefe** [comment]
> Made more progress on the review today, tomorrow I have the parser to review and java tests...

Thanks for your review work! I'll wait for your full review before continuing.

File: src/hotspot/share/runtime/os.cpp
Comment (gerard-ziemski) [reply src/hotspot/share/runtime/os.cpp:707]: The entire point of this effort is to catch elusive OOM errors. You even moved the check for the memory requirements before we actually request it, from the current way where we check the limit only after we acquire new memory.

Isn't assuming the worst case scenario, where realloc(), might allocate new memory first, in the same vein as that and we are just being cautious, to catch those hard to reproduce bugs?

This cases might not occur very often, granted, but why not catch them if we can?

Aren't there any platforms where we run Java where realloc() is implemented in terms of malloc() and memcpy()?

File: src/hotspot/share/services/mallocTracker.inline.hpp
Comment (gerard-ziemski) [reply src/hotspot/share/services/mallocTracker.inline.hpp:35]: That explains why we want to call `VMError::is_error_reported()` sure, but not why we wrapped it in `suppress_limit_handling()` SPI, unless I am missing something?

File: src/hotspot/share/services/mallocTracker.inline.hpp
Comment (gerard-ziemski) [reply src/hotspot/share/services/mallocTracker.inline.hpp:56]: I'm not following.

If we get rid of `suppress_limit_handling()` SPI, and do:

```
bool MallocMemorySummary::total_limit_reached(size_t s, size_t so_far, const malloclimit* limit) {
  if (VMError::is_error_reported()) {
    return false;
  }

#define FORMATTED \
"MallocLimit: reached global limit (triggering allocation size: " PROPERFMT ", allocated so far: " PROPERFMT ", limit: " PROPERFMT ") ", \
PROPERFMTARGS(s), PROPERFMTARGS(so_far), PROPERFMTARGS(limit->sz)

  if (limit->mode == MallocLimitMode::trigger_fatal) {
    fatal(FORMATTED);
  } else {
    log_warning(nmt)(FORMATTED);
  }

#undef FORMATTED

  return true;
}

bool MallocMemorySummary::category_limit_reached(MEMFLAGS f, size_t s, size_t so_far, const malloclimit* limit) {
  if (VMError::is_error_reported()) {
    return false;
  }

#define FORMATTED \
  "MallocLimit: reached category \"%s\" limit (triggering allocation size: " PROPERFMT ", allocated so far: " PROPERFMT ", limit: " PROPERFMT ") ", \
  NMTUtil::flag_to_enum_name(f), PROPERFMTARGS(s), PROPERFMTARGS(so_far), PROPERFMTARGS(limit->sz)

  if (limit->mode == MallocLimitMode::trigger_fatal) {
    fatal(FORMATTED);
  } else {
    log_warning(nmt)(FORMATTED);
  }

#undef FORMATTED

  return true;
}

```

then we can simply have:

```
// Returns true if allocating s bytes on f would trigger either global or the category limit
inline bool MallocMemorySummary::check_exceeds_limit(size_t s, MEMFLAGS f) {

  // Note: checks are ordered to have as little impact as possible on the standard code path,
  // when MallocLimit is unset, resp. it is set but we have reached no limit yet.
  // Somewhat expensive are:
  // - as_snapshot()->total(), total malloc load (requires iteration over arena types)
  // - VMError::is_error_reported() is a load from a volatile.
  if (MallocLimitHandler::have_limit()) {

    // Global Limit ?
    const malloclimit* l = MallocLimitHandler::global_limit();
    if (l->sz > 0) {
      size_t so_far = as_snapshot()->total();
      if ((so_far + s) > l->sz) { // hit the limit
        if (total_limit_reached(s, so_far, l)) {
          return true;
        }
      }
    } else {
      // Category Limit ?
      l = MallocLimitHandler::category_limit(f);
      if (l->sz > 0) {
        const MallocMemory* mm = as_snapshot()->by_type(f);
        size_t so_far = mm->malloc_size() + mm->arena_size();
        if ((so_far + s) > l->sz) {
          if (category_limit_reached(f, s, so_far, l)) {
            return true;
          }
        }
      }
    }
  }

  return false;
}

```

Doesn't that work?

File: src/hotspot/share/services/mallocLimit.cpp
Comment (gerard-ziemski) [inline src/hotspot/share/services/mallocLimit.cpp:150]: I am not crazy about:

`static const char* flagnames[] = { "fatal", "oom" };`

which requires that we position the strings in `flagnames` in a way that matches `MallocLimitMode` enum values:

```
enum class MallocLimitMode {
  trigger_fatal = 0,
  trigger_oom   = 1
};
```

Could we do instead:

```
void MallocLimitSet::print_on(outputStream* st) const {
  if (_glob.sz > 0) {
    st->print_cr("MallocLimit: total limit: " PROPERFMT " (%s)",
                 PROPERFMTARGS(_glob.sz), MallocLimitSet::limit_name(_glob.mode));
  } else {
    for (int i = 0; i < mt_number_of_types; i++) {
      if (_cat[i].sz > 0) {
        st->print_cr("MallocLimit: category \"%s\" limit: " PROPERFMT " (%s)",
          NMTUtil::flag_to_enum_name(NMTUtil::index_to_flag(i)),
          PROPERFMTARGS(_cat[i].sz), MallocLimitSet::limit_name(_cat[i].mode));
      }
    }
  }
}

```
and add new API to `MallocLimitSet`:

`  static const char* const limit_name(MallocLimitMode l)  { return (l==MallocLimitMode::trigger_fatal)?"fatal":"oom"; }
`

This way we also avoid the `(int)` casting from the `enum`.

We could also potentially use `MallocLimitSet::limit_name()` in `match_mode_flag()` instead of hardcoding "oom" and "fatal"

**Review by gerard-ziemski — CHANGES_REQUESTED** [review]
More feedback...

File: src/hotspot/share/runtime/os.cpp
Comment (tstuefe) [reply src/hotspot/share/runtime/os.cpp:707]: > The entire point of this effort is to catch elusive OOM errors. You even moved the check for the memory limits before we actually allocate the memory, from the current way where we check the limit only after we acquire new memory.
> 
> Isn't assuming the worst case scenario, where realloc(), might allocate new memory first, in the same vein as that and we are just being cautious, to catch those hard to reproduce bugs?
> 
> Basically we are saying that if the OS had to allocate new memory on top of the old one and doing so would exceed the limit we would like to know about it. Will it happen often? No? Will it happen during the lifetime of a memory intensive process? Most likely yes.
> 
> These cases might not occur very often, granted, but why not catch them if we can?
> 
> Aren't there any platforms where we run Java where realloc() is implemented in terms of malloc() and memcpy()?

I think its just wrong. False positives are not helpful.

1) its semantically wrong. The malloc limit shall triggers when our global or category-specific malloc load exceeds a given threshold. Be it either the malloc that just happened (old version) or the malloc that we are about to do (new version). But I don't want the mechanism to second-guess what I'm telling it and proactively trigger alerts. Example: I set 10m as limit and have a 5MB buffer. Expanding this buffer to 5.5 MB should not trigger any alert. I don't want false positives, they are not helpful.

2) You try to predict an OOM kill caused by a malloc event, right? For that to work the prediction must be at least somewhat correct. But you only have very limited knowledge. Some examples:
- The allocation request is fulfilled from cached memory. No new memory needed at all. The smaller the allocation the more likely this is, and if the libc uses sbrk exclusively, this is very likely.
- The libc always has overhead per allocation. E.g. glibc, at the very least, ~40ish bytes. Should I account for this? With many fine-grained allocations this can have an enourmous impact. But overhead depends on a number of factors, and every libc is different.
- The allocation request is large and the libc decides to mmap it (not all do). So we allocate in page granularity. Should I now adjust my prediction for page size? What page size? The libc may still reuse already mapped memory though. Or the underlying kernel vm may do that.
- The realloc is done by a new thread, and the libc uses thread-specific memory pools. A new memory pool is created. Is this memory committed immediately (e.g. older glibcs) or committed on demand? Depending on the answer, we may see a large bump in RSS or none at all. Depends on libc and libc version.
- The realloc is large and cannot be enlarged in place. But I could think of several ways of enlarging such an allocation out-of-place without needing memory * 2. If even I can think of those, libc devs probably do too.

So, these examples show that an allocation may not move the needle at all or may move the needle much farther than you think. Trying to guess this just makes the code complex and causes false positives that are difficult to argue about. All without having any real benefit.

Comment (gerard-ziemski) [reply src/hotspot/share/runtime/os.cpp:707]: Thank you for taking the time to explain.

I'm worrying about those hard to pin mysterious issues that we have plenty of that possibly could be related to memory pressure, but I understand your point of view and will defer to you (even if there is a 5% of this that continues to bother me).

File: src/hotspot/share/runtime/globals.hpp
Comment (gerard-ziemski) [reply src/hotspot/share/runtime/globals.hpp:1344]: Still, you named the test file `test/hotspot/gtest/nmt/test_nmt_malloclimit.cpp`, not `test/hotspot/gtest/nmt/test_malloclimit.cpp` :-)

**Comment by gerard-ziemski** [comment]
The rest looks good to me.

This is a really nice feature, thank you for doing this!

File: src/hotspot/share/runtime/os.cpp
Comment (tstuefe) [reply src/hotspot/share/runtime/os.cpp:707]: I understand. This is not the tool to solve all that, but it can help. 

There are other tools. At SAP, we added something called "High Memory Reports", the ability to automatically generate customizable reports when certain memory threshold is reached [1]. In addition to that, we have our SapMachine Vitals [2], which are awesome  - basically like a little inbuilt pidstat but with a lot more information. For instance, it tells you the history of c-heap allocations (both from the JVM view and from the libc's view).

Would love to see this in OpenJDK, but have zero hopes of doing so - tried to get the Vitals in years ago but hit a wall of resistance from JFR people.

[1] https://github.com/SAP/SapMachine/wiki/SapMachine-High-Memory-Reports
[2] https://github.com/SAP/SapMachine/wiki/SapMachine-Vitals

File: src/hotspot/share/runtime/globals.hpp
Comment (tstuefe) [reply src/hotspot/share/runtime/globals.hpp:1344]: I'm not perfect :-) honestly, I just like short, easy to remember and to talk about command line names. I dislike the JVM tradition of long camel-cased names a bit.

File: src/hotspot/share/services/mallocTracker.inline.hpp
Comment (tstuefe) [reply src/hotspot/share/services/mallocTracker.inline.hpp:35]: Oh okay. This was mostly a documentation thing. I can remove the wrapper.

File: src/hotspot/share/services/mallocTracker.inline.hpp
Comment (tstuefe) [reply src/hotspot/share/services/mallocTracker.inline.hpp:56]: Yes, you are right, its better. Can remove vmError.hpp dependency from the header too.

File: src/hotspot/share/services/mallocLimit.cpp
Comment (tstuefe) [reply src/hotspot/share/services/mallocLimit.cpp]: Okay, minus the exposure of the API. Its not needed outside of this file.

**Comment by tstuefe** [comment]
Hi @gerard-ziemski, thanks for your review! Could you take a look at the final version?

File: src/hotspot/share/services/mallocLimit.cpp
Comment (gerard-ziemski) [inline src/hotspot/share/services/mallocLimit.cpp:149]: This still requires us to be careful and match the order in

enum class MallocLimitMode {
  trigger_fatal = 0,
  trigger_oom   = 1
};

so I like my proposal better, but at least we're not hardcoding it here, so it's an improvement.

**Review by gerard-ziemski — APPROVED** [review]
Looks good!

**Comment by tstuefe** [comment]
Thanks, Gerard!

@jdksjolen are you fine with the updated version?

**Comment by tstuefe** [comment]
@jdksjolen are you fine with the updated version? I would like to integrate this. 

This has been cooking in our test systems for about 3 months now, so I think it is well-tested.

**Review by jdksjolen — APPROVED** [review]
Hi @tstuefe , thanks for the extra ping. Yes, this looks good!

**Comment by jdksjolen** [comment]
Sorry, could you please convert the NULLs into nullptr :)?

**Comment by tstuefe** [comment]
> Sorry, could you please convert the NULLs into nullptr :)?

Done. Thanks for the review; I'll wait for the final GHAs, then I'll push

**Comment by tstuefe** [comment]
@jdksjolen @gerard-ziemski could you please re-approve? For some reason I cannot integrate and I suspect its because reviews are stale.

**Review by jdksjolen — APPROVED** [review]
Re-approving!

**Review by RealLucy — APPROVED** [review]
LGTM.
Thanks for the work and relentless discussion.

**Review by MBaesken — APPROVED** [review]
LGTM.

Will there be a release note documenting the changed/removed XX-flags ? Maybe there should be one .

**Comment by dholmes-ora** [comment]
@tstuefe with regards to a release note ... while the flag is now diagnostic that is a change made in JDK 21, so in terms of users of JDK 20 migrating to JDK 21, a release note may be  need to both explain the flag is now diagnostic (and subject to change without notice), and that "none" is no longer a valid value.

**Comment by tstuefe** [comment]
> @tstuefe with regards to a release note ... while the flag is now diagnostic that is a change made in JDK 21, so in terms of users of JDK 20 migrating to JDK 21, a release note may be need to both explain the flag is now diagnostic (and subject to change without notice), and that "none" is no longer a valid value.

I assume you mean JDK-8302129, right? The MetaspaceReclaimPolicy=none removal? 

Makes sense, I will do that. I opened https://bugs.openjdk.org/browse/JDK-8302708 to track that.

**Comment by dholmes-ora** [comment]
Sorry @tstuefe - no idea how I got the two issues mixed up. :(

**Pull request closed without merging**
